File: vertx-web-mongo/src/main/java/net/aotter/vertx_web_mongo/MainVerticle.java
package net.aotter.vertx_web_mongo;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.TimeZone;

import io.netty.handler.codec.http.HttpContent;
import io.vertx.core.AbstractVerticle;
import io.vertx.core.Future;
import io.vertx.core.Promise;
import io.vertx.core.http.HttpServer;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.mongo.FindOptions;
import io.vertx.ext.mongo.MongoClient;
import io.vertx.ext.mongo.UpdateOptions;
import io.vertx.ext.web.Router;
import io.vertx.ext.web.RoutingContext;
import org.apache.commons.lang3.time.DateUtils;

public class MainVerticle extends AbstractVerticle {


  private MongoClient mongoClient;


  private final DateFormat dateFormat = new SimpleDateFormat(
    "yyyy-MM-dd'T'HH:mm'Z'");


  public MainVerticle() {
    dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
  }


  /**
   * app init here
   * @param startPromise
   * @throws Exception
   */
  @Override
  public void start(Promise<Void> startPromise) throws Exception {

    mongoClient = MongoClient.createShared(vertx,
      new JsonObject()
        .put("connection_string", "mongodb://localhost:27017")
        .put("db_name", "reactive-load-test")
    );

    HttpServer server = vertx.createHttpServer();
    Router router = Router.router(vertx);

    router.get("/report").produces("text/plain").handler(this::report);

    server.requestHandler(router).listen(8080, http -> {
      if (http.succeeded()) {
        startPromise.complete();
        System.out.println("HTTP server started on port 8080");
      } else {
        startPromise.fail(http.cause());
      }
    });
  }


  /**
   * request handler
   * @param ctx
   */
  private void report(RoutingContext ctx) {
    upsert().onComplete(res -> {
      if (res.succeeded()) {
        Long count = res.result().getLong("count");
        ctx.response().end(String.format("#%d", count));
      } else {
        res.cause().printStackTrace();
        ctx.response().setStatusCode(500).end(res.cause().getMessage());
      }
    });
  }


  /**
   * upsert into mongoDB
   *
   * @return
   */
  private Future<JsonObject> upsert() {
    // Note: remember to add unique index to "hour" field
    return mongoClient.findOneAndUpdateWithOptions(
      "VertxWebMongo",
      new JsonObject().put("date",
        new JsonObject().put("$date", dateFormat.format(DateUtils.truncate(new Date(), Calendar.HOUR)))),
      new JsonObject().put("$inc", new JsonObject().put("count", 1L)),
      new FindOptions(),
      new UpdateOptions().setUpsert(true).setReturningNewDocument(true)
    );
  }

}

File: vertx-web-kotlin-coroutine-mongo/src/main/kotlin/net/aotter/vertx_web_kotlin_coroutine_mongo/MainVerticle.kt
package net.aotter.vertx_web_kotlin_coroutine_mongo

import io.vertx.core.json.JsonObject
import io.vertx.ext.mongo.FindOptions
import io.vertx.ext.mongo.MongoClient
import io.vertx.ext.mongo.UpdateOptions
import io.vertx.ext.web.Router
import io.vertx.kotlin.core.json.jsonObjectOf
import io.vertx.kotlin.coroutines.CoroutineVerticle
import io.vertx.kotlin.coroutines.await
import net.aotter.vertx_web_kotlin_coroutine_mongo.extension.coroutineHandler
import org.apache.commons.lang3.time.DateUtils
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.*

class MainVerticle : CoroutineVerticle() {

  lateinit var mongoClient: MongoClient

  private val dateFormat: DateFormat = SimpleDateFormat(
    "yyyy-MM-dd'T'HH:mm'Z'"
  )

  init {
    dateFormat.timeZone = TimeZone.getTimeZone("UTC")
  }


  /**
   * app init
   */
  override suspend fun start() {

    // connect to mongoDB
    mongoClient = MongoClient.createShared(
      vertx,
      jsonObjectOf(
        "connection_string" to "mongodb://localhost:27017",
        "db_name" to "reactive-load-test"
      )
    )

    // declare route
    val router = Router.router(vertx)
    router.get("/report").produces("text/plain").coroutineHandler {
      val result = upsert()
      it.response().end("#${result?.getLong("count") ?: "N/A"}")
    }

    // init server
    vertx.createHttpServer().requestHandler(router)
      .listen(8080) { println("HTTP server started on port 8080 succeeded: ${it.succeeded()}") }
  }


  /**
   * upsert mongo
   * Note: remember to add unique index to "hour" field
   */
  private suspend fun upsert(): JsonObject? = mongoClient.findOneAndUpdateWithOptions(
    "VertxWebKotlinCoroutineMongo",
    jsonObjectOf("date" to jsonObjectOf("\$date" to Date().toStartOfCurrentHour().formatIsoDate())),
    jsonObjectOf("\$inc" to jsonObjectOf("count" to 1L)),
    FindOptions(),
    UpdateOptions().setUpsert(true).setReturningNewDocument(true)
  ).await()


  private fun Date.toStartOfCurrentHour(): Date = DateUtils.truncate(this, Calendar.HOUR)


  private fun Date.formatIsoDate(): String = dateFormat.format(this)

}

